module/zfs/arc.c

```diff
@@ -3922,9 +3922,11 @@ arc_evict_hdr(arc_buf_hdr_t *hdr, kmutex_t *hash_lock)
 			 */
 			hdr = arc_hdr_realloc(hdr, hdr_full_cache,
 			    hdr_l2only_cache);
+			*real_evicted += HDR_FULL_SIZE - HDR_L2ONLY_SIZE;
 		} else {
 			arc_change_state(arc_anon, hdr, hash_lock);
 			arc_hdr_destroy(hdr);
+			*real_evicted += HDR_FULL_SIZE;
```

Comment:
I think it is already consistent as much as it can get.  Evictions from non-ghost states always go to ghost states and so don't free headers, so there is nothing more to account in bytes_evicted.  Evictions from ghost states use bytes_evicted to make ghost eviction go the same rate as non-ghost in terms of data size, and since the last does not account for headers, the second also shouldn't.  Any way the difference is far below the noise level.

Reply:
Oh yeah, I forgot that the hdr is not actually freed in the non-ghost case.  So bytes_evicted and real_evicted should not include the hdr size in the non-ghost case, and after re-reading the code they do not (arc_hdr_size() is the size of the data in the hdr, not the header itself). 

If we want to define real_evicted as a subset of bytes_evicted, then in the ghost case we should add the hdr size to both real_evicted and bytes_evicted.  Alternatively, we could define bytes_evicted as being just the data buffer size (or "virtual size" for ghost), and real_evicted as being the amount of memory actually freed (hdr + actual data).  The latter seems to be what's implemented here.

This all seems fine, but it would be helpful to have some comments explaining the real_evicted and bytes_evicted quantities.

Reply:
I've added comments about it.